### PR TITLE
HW#6 and HW#7

### PR DESCRIPTION
**Pull request by 1dynamite**
I have added login page, client-side validations, error-handling, interceptors, guards, etc

P.S. I still have 30 minutes before it hits Monday, the deadline. So, I have submitted the HWs in time.

### CODE REVIEW

File: client/src/app/dashboard/dashboard.component.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import { DashboardService } from './dashboard.service';
+
+@Component({
+  selector: 'app-dashboard',
+  templateUrl: './dashboard.component.html',
+  styleUrls: ['./dashboard.component.scss'],
+})
+export class DashboardComponent implements OnInit {
+  user: any;
```
Comment (AudriusBaltusis) [inline client/src/app/dashboard/dashboard.component.ts:12]: Please do not use `any`

File: client/src/app/dashboard/dashboard.component.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      error: (err: any) => {
+        console.error(err);
+        this.router.navigateByUrl('/login');
+      },
+    };
+
+    if (!this.user) this.dashboardService.getUser().subscribe(myObserver);
+  }
+
+  logout() {
```
Comment (AudriusBaltusis) [inline client/src/app/dashboard/dashboard.component.ts:35]: Please add return type

File: client/src/app/dashboard/dashboard.service.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    if (token) {
+      const decoded: any = jwt_decode(token);
+
+      this.userUrl = `/api/users/${decoded._id}`;
+    } else {
+      this.userUrl = null;
+    }
+  }
+
+  getUser() {
```
Comment (AudriusBaltusis) [inline client/src/app/dashboard/dashboard.service.ts:27]: Please add return type everywhere

File: client/src/app/login/login.service.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+export class LoginService {
+  private loginUrl = '/api/login';
+
+  httpOptions = {
+    headers: new HttpHeaders({ 'Content-Type': 'application/json' }),
+  };
+
+  constructor(private http: HttpClient) {}
+
+  login(loginData: LoginData): Observable<any> {
```
Comment (AudriusBaltusis) [inline client/src/app/login/login.service.ts:23]: Please do no use `any`

File: client/src/app/not-found/not-found.component.html [hunk 5]
```diff
@@ -0,0 +1 @@
+<p>Page not found(</p>
```
Comment (AudriusBaltusis) [inline client/src/app/not-found/not-found.component.html:1]: Typo `(`

File: client/src/app/not-found/not-found.component.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+@Component({
+  selector: 'app-not-found',
+  templateUrl: './not-found.component.html',
+  styleUrls: ['./not-found.component.scss']
+})
+export class NotFoundComponent implements OnInit {
+
+  constructor() { }
+
+  ngOnInit(): void {
```
Comment (AudriusBaltusis) [inline client/src/app/not-found/not-found.component.ts:12]: Please remove empty `ngOnInit`

**Review by AudriusBaltusis — CHANGES_REQUESTED** [review]
If this PR contains HW #7, you have to create components for header and footer

**Comment by 1dynamite** [comment]
I am developing the UI based on the Figma design provided by Maxim (https://www.figma.com/file/EXv9NCHklgfbbhB4FfgaS0/Budgetify?node-id=0%3A1). This design does not have a footer and one big header. Instead, it is divided into three main columns. That's what I did with my dashboard component. But if you insist that I should have a footer and a header, then I will create them. What do you think?

**Comment by AudriusBaltusis** [comment]
> I am developing the UI based on the Figma design provided by Maxim (https://www.figma.com/file/EXv9NCHklgfbbhB4FfgaS0/Budgetify?node-id=0%3A1). This design does not have a footer and one big header. Instead, it is divided into three main columns. That's what I did with my dashboard component. But if you insist that I should have a footer and a header, then I will create them. What do you think?

If there is no footer in design, then skip it. But header component is mandatory

**Comment by 1dynamite** [comment]
I do have a header. 
image

**Comment by 1dynamite** [comment]
Ok, I think I got you. I will extract those top-level bars (or headers) from the dashboard component and put them in separate files and import them.

**Comment by 1dynamite** [comment]
image

**Review by AudriusBaltusis — APPROVED** [review]

**Pull request merged**